gprofiler/profilers/node.py

```diff
@@ -0,0 +1,244 @@
+import json
+import os
+import shutil
+import signal
+import stat
+from functools import lru_cache
+from threading import Event
+from typing import Dict, List, cast
+
+import psutil
+import requests
+from granulate_utils.linux.ns import get_proc_root_path, get_process_nspid, resolve_proc_root_links, run_in_ns
+from granulate_utils.linux.process import is_musl, is_process_running
+from retry import retry
+from websocket import create_connection
+from websocket._core import WebSocket
+
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata.versions import get_exe_version
+from gprofiler.utils import TEMPORARY_STORAGE_PATH, add_permission_dir, pgrep_exe, resource_path
+
+logger = get_logger_adapter(__name__)
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def _get_node_major_version(process: psutil.Process) -> str:
+    node_version = get_exe_version(process, Event(), 3)
+    # i. e. v16.3.2 -> 16
+    return node_version[1:].split(".")[0]
+
+
+@lru_cache(maxsize=1)
+def _get_dso_git_rev() -> str:
+    libc_dso_version_file = resource_path(os.path.join("node", "module", "glibc", "version"))
+    musl_dso_version_file = resource_path(os.path.join("node", "module", "musl", "version"))
+    with open(libc_dso_version_file) as f:
+        libc_dso_ver = f.readline()
+    with open(musl_dso_version_file) as f:
+        musl_dso_ver = f.readline()
+    # with no build errors this should always be the same
+    assert libc_dso_ver == musl_dso_ver
+    return libc_dso_ver
+
+
+@lru_cache()
+def _get_dest_inside_container(musl: bool, node_version: str) -> str:
+    libc = "musl" if musl else "glibc"
+    return os.path.join(TEMPORARY_STORAGE_PATH, "node_module", _get_dso_git_rev(), libc, node_version)
+
+
+def _start_debugger(pid: int) -> None:
+    # for windows: in shell node -e "process._debugProcess(PID)"
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same. When another debugger is opened in same NS it will not open new one.
+    # REF: Inspector agent initialization uses host_port
+    # https://github.com/nodejs/node/blob/5fad0b93667ffc6e4def52996b9529ac99b26319/src/inspector_agent.cc#L668
+    # host_port defaults to 9229
+    # ref: https://github.com/nodejs/node/blob/2849283c4cebbfbf523cc24303941dc36df9332f/src/node_options.h#L90
+    # in our case it won't be changed
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
```

Comment:
Let's add timeouts to all requests to prevent the profiler from being stuck if any process misbehaves. I think 3 seconds would suffice?

Reply:
In `create_debugger_socket` we already have a timeout, so only for this `requests.get` here it would suffice I suppose